### PR TITLE
Geomap: Support geojson styling properties

### LINKED ISSUE #76083: Geomap: geojson styling
**Why is this needed**:

Styling of geojson boundaries

**What would you like to be added**:

ability to use css styles in geojson

**Who is this feature for?**

The World! ;)

I notice this specification for some mapping utilities, not sure if this feature is available in geomap?

https://wiki.openstreetmap.org/wiki/Geojson_CSS

```
{ 
    "type": "Feature",
    "geometry": {
    "type": "Polygon",
    "coordinates": [[
        [-180.0, 10.0], [20.0, 90.0], [180.0, -5.0], [-30.0, -90.0]
        ]]
    },
    "style": {
        "__comment": "all SVG styles allowed",
        "fill":"red",
        "stroke-width":"3",
        "fill-opacity":0.6
    },
    "className": {
        "baseVal":"A class name"
    }
}
```

### PR DESCRIPTION
**Pull request by drew08t**
Add support for embedding polygon styles (fill, fill-opacity, stroke-width) within geojson feature properties. Based on [community contribution](https://github.com/grafana/grafana/pull/79587).

Based on this [geojson styling spec](https://github.com/mapbox/simplestyle-spec/tree/master/1.1.0) which conforms to [geojson spec](https://datatracker.ietf.org/doc/html/rfc7946#section-1.5), specifically for feature properties.

A geojson with embedded styles has been included as an example.

- [x] Setup enum for property types supported
- [x] Docs update
- [x] Update example to be simpler
- [x] Point https://github.com/grafana/grafana/issues/80274
- [x] MultiPoint https://github.com/grafana/grafana/issues/80491
- [x] LineString
- [x] MultiLineString
- [x] Polygon
- [x] MultiPolygon

image

Closes #76083, #80491, #80274

### CODE REVIEW

File: public/app/plugins/panel/geomap/layers/data/geojsonLayer.ts [hunk 1]
```diff
@@ -131,11 +125,24 @@ export const geojsonLayer: MapLayerRegistryItem<GeoJSONMapperConfig> = {
             return polyStyle(values);
           }
 
+          // Support styling polygons from Feature properties
+          const featureProps = feature.getProperties();
+          const styleStrings: string[] = Object.values(GeoJSONStyles);
```
Comment (ryantxu) [inline public/app/plugins/panel/geomap/layers/data/geojsonLayer.ts:130]: move this outside the inner loop -- currently it will be created for each feature

File: public/maps/time-zones-styled.geojson
Comment (ryantxu) [inline public/maps/time-zones-styled.geojson:1]: is this worth adding to the distribution?  800k for this kinda essoteric example seems not worth it.  Perhaps a smaller test example?  that exercises points, lines, and polygons?

File: public/app/plugins/panel/geomap/layers/data/geojsonLayer.ts (hunk 1 above)
Comment (drew08t) [reply public/app/plugins/panel/geomap/layers/data/geojsonLayer.ts:130]: Good catch!

File: public/maps/time-zones-styled.geojson
Comment (drew08t) [reply public/maps/time-zones-styled.geojson:1]: Adding a simpler example dedicated as a style example.

File: docs/sources/panels-visualizations/visualizations/geomap/index.md [hunk 2]
```diff
@@ -282,6 +282,16 @@ The GeoJSON layer allows you to select and load a static GeoJSON file from the f
 - **Add style rule** creates additional style rules.
 - **Display tooltip** allows you to toggle tooltips for the layer.
 
+{{% admonition type="note" %}}
```
Comment (baldm0mma) [inline docs/sources/panels-visualizations/visualizations/geomap/index.md:285]: I think we're no longer using Hugo `admonition` short codes in favor of `> **Note:** <you're note here>` syntax. @imatwawana can you confirm?

Comment (imatwawana) [reply docs/sources/panels-visualizations/visualizations/geomap/index.md:285]: We favour admonition shortcodes, but in the `transformations.md` file we can't use them because of how it's generated, so that might be what's causing confusion here.

File: docs/sources/panels-visualizations/visualizations/geomap/index.md [hunk 3]
```diff
@@ -282,6 +282,16 @@ The GeoJSON layer allows you to select and load a static GeoJSON file from the f
...
+Styles can be set within the "properties" object of the GeoJSON with support for the following geometrys.
```
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/geomap/index.md:286]:
```suggestion
Styles can be set within the `properties` object of the GeoJSON with support for the following geometries:
```

File: docs/sources/panels-visualizations/visualizations/geomap/index.md [hunk 4]
```diff
@@ -282,6 +282,16 @@ The GeoJSON layer allows you to select and load a static GeoJSON file from the f
...
+
+- Polygon, MultiPolygon
+  - **"fill"** the color of the interior of the polygon
+  - **"fill-opacity"** the opacity of the interior of the polygon
+  - **"stroke-width"** the width of the line component of the polygon
```
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/geomap/index.md:291]: Do we need both the quotation marks and the bold? If these are properties in the JSON object, I would suggest the following formatting: 
```suggestion
  - `fill` - The color of the interior of the polygon
  - `fill-opacity` - The opacity of the interior of the polygon
  - `stroke-width` - The width of the line component of the polygon
```

**Review by imatwawana — COMMENTED** [review]
Left some formatting suggestions

File: docs/sources/panels-visualizations/visualizations/geomap/index.md (hunk 3 above)
Comment (drew08t) [reply docs/sources/panels-visualizations/visualizations/geomap/index.md:286]: Good suggestions, but will keep double quotes which matches the format of the json standard.

File: docs/sources/panels-visualizations/visualizations/geomap/index.md (hunk 4 above)
Comment (drew08t) [reply docs/sources/panels-visualizations/visualizations/geomap/index.md]: Adding the hyphen and capitalization, but keeping the double quotes to be consistent with the geojson standard.

**Review by nmarrs — APPROVED** [review]
tested locally and this is looking good!

**Comment by nmarrs** [comment]
The options seem to be off based on the current color 🤔 not sure what is going on there 

Screenshot 2024-01-12 at 8 34 09 PM

File: public/app/plugins/panel/geomap/layers/data/geojsonLayer.ts [hunk 5]
```diff
@@ -131,6 +139,29 @@ export const geojsonLayer: MapLayerRegistryItem<GeoJSONMapperConfig> = {
             return polyStyle(values);
           }
 
+          // Support styling polygons from Feature properties
+          const featureProps = feature.getProperties();
+          if (isPolygon && Object.keys(featureProps).some((property) => polyStyleStrings.includes(property))) {
```
Comment (codeincarnate) [inline public/app/plugins/panel/geomap/layers/data/geojsonLayer.ts:144]: I'm a bit concerned about the property comparisons here. Seems like there could be a situation in which we have M `featureProps` along with N style strings resulting in worse case M*N in terms of comparisons. Probably not meaningful in small layers, but I could see this being hairy in larger ones potentially.

**Review by codeincarnate — APPROVED** [review]
Tested this out locally and looks good! I'm a bit concerned by the comparisons but beyond that seems fine 😄

File: public/app/plugins/panel/geomap/layers/data/geojsonLayer.ts (hunk 5 above)
Comment (drew08t) [reply public/app/plugins/panel/geomap/layers/data/geojsonLayer.ts:144]: Thanks for testing it!

Perhaps we could do the embedded styling check for geojson on first load and store a file content hash with hasStyling boolean inside of the panel json. That way, if there is no embedded styling in a particular source (that contents hasn't changed), it avoids these loops altogether. Of course doing the hash check wouldn't be free and there may be a crossover point where it doesn't make sense. For the case where embedded styling is found, we could store a comprehensive map of which styles were used to reduce these loops.

TLDR: we should follow-up on how to better optimize this.

**Comment by drew08t** [comment]
> The options seem to be off based on the current color 🤔 not sure what is going on there
> 
> Screenshot 2024-01-12 at 8 34 09 PM

Hmm, that example definitely shouldn't look like that. Are you still experiencing this after panel refresh?

File: public/app/plugins/panel/geomap/style/types.ts [hunk 6]
```diff
@@ -127,6 +127,22 @@ export interface StyleConfigValues {
   textConfig?: TextStyleConfig;
 }
 
+export enum GeoJSONPolyStyles {
```
Comment (baldm0mma) [inline public/app/plugins/panel/geomap/style/types.ts:130]: Nit: most other places in our codebase I think we capitalize `enum` properties.

Comment (drew08t) [reply public/app/plugins/panel/geomap/style/types.ts:130]: Good point. I wanted it to match the [StyleConfigValues properties](https://github.com/grafana/grafana/blob/18b9c8fd5fda882013e5224e48ce029d8259aee7/public/app/plugins/panel/geomap/style/types.ts#L116), but I suppose there may be a cleaner way to approach this.

File: public/app/plugins/panel/geomap/layers/data/geojsonLayer.ts (hunk 5 above)
Comment (codeincarnate) [reply public/app/plugins/panel/geomap/layers/data/geojsonLayer.ts:144]: That sounds like a really good way forward! Really appreciate the thoughtful approach here 😄

**Comment by nmarrs** [comment]
Internal note: add this to misc viz improvements for 10.4

**Comment by yosiasz** [comment]
is the polygon feature now in 11? @drew08t thanks
